File: ta4j-core/src/main/java/org/ta4j/core/rules/StopLossRule.java
/*
 * SPDX-License-Identifier: MIT
 */
package org.ta4j.core.rules;

import java.util.Objects;

import org.ta4j.core.BarSeries;
import org.ta4j.core.Indicator;
import org.ta4j.core.Position;
import org.ta4j.core.TradingRecord;
import org.ta4j.core.num.Num;

/**
 * A stop-loss rule.
 *
 * <p>
 * Satisfied when the configured price indicator touches or moves beyond the
 * loss threshold. Long positions trigger at or below the stop price; short
 * positions trigger at or above it.
 *
 * <p>
 * This rule uses the {@code tradingRecord}.
 */
public class StopLossRule extends AbstractRule implements StopLossPriceModel {

    /** The reference price indicator. */
    private final Indicator<Num> priceIndicator;

    /** The loss percentage. */
    private final Num lossPercentage;

    /**
     * Constructor.
     *
     * @param priceIndicator the price indicator
     * @param lossPercentage the loss percentage
     */
    public StopLossRule(Indicator<Num> priceIndicator, Number lossPercentage) {
        this(priceIndicator, numOf(priceIndicator, lossPercentage, "lossPercentage"));
    }

    /**
     * Constructor.
     *
     * @param priceIndicator the price indicator
     * @param lossPercentage the loss percentage
     */
    public StopLossRule(Indicator<Num> priceIndicator, Num lossPercentage) {
        this.priceIndicator = Objects.requireNonNull(priceIndicator, "priceIndicator");
        if (Num.isNaNOrNull(lossPercentage) || lossPercentage.isNegative()) {
            throw new IllegalArgumentException("lossPercentage must be >= 0");
        }
        this.lossPercentage = lossPercentage;
    }

    /**
     * Computes the stop-loss price from the entry price and loss percentage.
     *
     * @param entryPrice     the entry price
     * @param lossPercentage the loss percentage
     * @param isBuy          true for long positions, false for short positions
     * @return the stop-loss price
     * @since 0.22.3
     */
    public static Num stopLossPrice(Num entryPrice, Num lossPercentage, boolean isBuy) {
        if (Num.isNaNOrNull(entryPrice)) {
            throw new IllegalArgumentException("entryPrice must not be null");
        }
        if (Num.isNaNOrNull(lossPercentage) || lossPercentage.isNegative()) {
            throw new IllegalArgumentException("lossPercentage must be >= 0");
        }
        Num hundred = entryPrice.getNumFactory().hundred();
        Num lossRatioThreshold = isBuy ? hundred.minus(lossPercentage).dividedBy(hundred)
                : hundred.plus(lossPercentage).dividedBy(hundred);
        return entryPrice.multipliedBy(lossRatioThreshold);
    }

    /**
     * Computes the stop-loss price from the entry price and an absolute loss
     * distance.
     *
     * @param entryPrice   the entry price
     * @param lossDistance the absolute price distance to the stop
     * @param isBuy        true for long positions, false for short positions
     * @return the stop-loss price
     * @since 0.22.3
     */
    public static Num stopLossPriceFromDistance(Num entryPrice, Num lossDistance, boolean isBuy) {
        if (Num.isNaNOrNull(entryPrice)) {
            throw new IllegalArgumentException("entryPrice must not be null");
        }
        if (Num.isNaNOrNull(lossDistance) || lossDistance.isNegative()) {
            throw new IllegalArgumentException("lossDistance must be >= 0");
        }
        return isBuy ? entryPrice.minus(lossDistance) : entryPrice.plus(lossDistance);
    }

    /**
     * Returns the stop-loss price for the supplied position entry.
     *
     * @param series   the price series (unused in this implementation; required by
     *                 {@link StopLossPriceModel})
     * @param position the position being evaluated
     * @return the stop-loss price, or {@code null} if unavailable
     * @since 0.22.3
     */
    @Override
    public Num stopPrice(BarSeries series, Position position) {
        if (position == null || position.getEntry() == null) {
            return null;
        }
        Num entryPrice = position.getEntry().getNetPrice();
        if (Num.isNaNOrNull(entryPrice)) {
            return null;
        }
        return stopLossPrice(entryPrice, lossPercentage, position.getEntry().isBuy());
    }

    /** This rule uses the {@code tradingRecord}. */
    @Override
    public boolean isSatisfied(int index, TradingRecord tradingRecord) {
        boolean satisfied = false;
        // No trading history or no position opened, no loss
        if (tradingRecord != null) {
            Position currentPosition = tradingRecord.getCurrentPosition();
            if (currentPosition != null && currentPosition.isOpened() && currentPosition.getEntry() != null) {

                Num entryPrice = currentPosition.getEntry().getNetPrice();
                Num currentPrice = priceIndicator.getValue(index);

                if (!Num.isNaNOrNull(entryPrice) && !Num.isNaNOrNull(currentPrice)
                        && currentPosition.getEntry().isBuy()) {
                    satisfied = isBuyStopSatisfied(entryPrice, currentPrice);
                } else if (!Num.isNaNOrNull(entryPrice) && !Num.isNaNOrNull(currentPrice)) {
                    satisfied = isSellStopSatisfied(entryPrice, currentPrice);
                }
            }
        }
        traceIsSatisfied(index, satisfied);
        return satisfied;
    }

    /**
     * Checks stop-loss trigger condition for a long position.
     *
     * @param entryPrice   entry price
     * @param currentPrice current price
     * @return {@code true} when current price reaches long stop-loss threshold
     */
    private boolean isBuyStopSatisfied(Num entryPrice, Num currentPrice) {
        Num threshold = stopLossPrice(entryPrice, lossPercentage, true);
        return currentPrice.isLessThanOrEqual(threshold);
    }

    /**
     * Checks stop-loss trigger condition for a short position.
     *
     * @param entryPrice   entry price
     * @param currentPrice current price
     * @return {@code true} when current price reaches short stop-loss threshold
     */
    private boolean isSellStopSatisfied(Num entryPrice, Num currentPrice) {
        Num threshold = stopLossPrice(entryPrice, lossPercentage, false);
        return currentPrice.isGreaterThanOrEqual(threshold);
    }

    private static Num numOf(Indicator<Num> priceIndicator, Number value, String parameterName) {
        Objects.requireNonNull(priceIndicator, "priceIndicator");
        Objects.requireNonNull(value, parameterName);
        return priceIndicator.getBarSeries().numFactory().numOf(value);
    }
}

File: ta4j-core/src/main/kotlin/org/ta4j/core/rules/StopLossRule.kt
/**
 * The MIT License (MIT)
 *
 * Copyright (c) 2017-2022 Ta4j Organization & respective
 * authors (see AUTHORS)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
 * the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
package org.ta4j.core.rules

import org.ta4j.core.TradingRecord
import org.ta4j.core.indicators.helpers.ClosePriceIndicator
import org.ta4j.core.num.*

/**
 * A stop-loss rule.
 *
 * Satisfied when the close price reaches the loss threshold.
 */
class StopLossRule(
    /**
     * The close price indicator
     */
    private val closePrice: ClosePriceIndicator,
    /**
     * The loss percentage
     */
    private val lossPercentage: Num
) : AbstractRule() {
    /**
     * Constant value for 100
     */
    private val HUNDRED: Num?

    /**
     * Constructor.
     *
     * @param closePrice     the close price indicator
     * @param lossPercentage the loss percentage
     */
    constructor(closePrice: ClosePriceIndicator, lossPercentage: Number) : this(
        closePrice,
        closePrice.numOf(lossPercentage)
    ) {
    }

    /**
     * Constructor.
     *
     * @param closePrice     the close price indicator
     * @param lossPercentage the loss percentage
     */
    init {
        HUNDRED = closePrice.numOf(100)
    }

    /** This rule uses the `tradingRecord`.  */
    override fun isSatisfied(index: Int, tradingRecord: TradingRecord?): Boolean {
        var satisfied = false
        // No trading history or no position opened, no loss
        if (tradingRecord != null) {
            val currentPosition = tradingRecord.getCurrentPosition()
            if (currentPosition.isOpened) {
                val entryPrice = currentPosition.entry!!.netPrice
                val currentPrice = closePrice[index]
                satisfied = if (currentPosition.entry!!.isBuy) {
                    isBuyStopSatisfied(entryPrice, currentPrice)
                } else {
                    isSellStopSatisfied(entryPrice, currentPrice)
                }
            }
        }
        traceIsSatisfied(index, satisfied)
        return satisfied
    }

    private fun isBuyStopSatisfied(entryPrice: Num?, currentPrice: Num?): Boolean {
        val lossRatioThreshold = HUNDRED!!.minus(lossPercentage).div(HUNDRED)
        val threshold = entryPrice!!.times(lossRatioThreshold)
        return currentPrice!!.isLessThanOrEqual(threshold)
    }

    private fun isSellStopSatisfied(entryPrice: Num?, currentPrice: Num?): Boolean {
        val lossRatioThreshold = HUNDRED!!.plus(lossPercentage).div(HUNDRED)
        val threshold = entryPrice!!.times(lossRatioThreshold)
        return currentPrice!!.isGreaterThanOrEqual(threshold)
    }
}